framework/contrib/hit/main.cc

```diff
@@ -4,132 +4,332 @@
+    _have[n->fullpath()] = n;
   }
-  std::map<std::string, std::vector<hit::Node *>> duplicates;
+
+  std::vector<std::string> errors;
 
 private:
-  std::map<std::string, hit::Node *> have;
+  std::string _fname;
+  std::set<std::string> _duplicates;
+  std::map<std::string, hit::Node *> _have;
 };
 
 int
-checkfiles(int argc, char ** argv)
+renameParam(int argc, char ** argv)
 {
+  Flags flags;
+  flags.add("i", "modify file(s) inplace");
+  auto positional = parseOpts(argc, argv, flags);
+
+  if (positional.size() < 3)
+  {
+    std::cerr << "rename expects 2 arguments: [src-path] [dst-path] [file]...";
+    return 1;
+  }
+
+  std::string src(positional[0]);
+  std::string dst(positional[1]);
+
   int ret = 0;
-  if (argc > 1)
+  for (int i = 2; i < positional.size(); i++)
   {
-    for (int i = 1; i < argc; i++)
-    {
-      std::string fname(argv[i]);
+    std::string fname(positional[i]);
+    std::ifstream f(fname);
+    std::string input((std::istreambuf_iterator<char>(f)), std::istreambuf_iterator<char>());
 
-      std::ifstream f(fname);
-      std::stringstream ss;
-      ss << f.rdbuf();
-      std::string input = ss.str();
+    hit::Node * root = nullptr;
+    try
+    {
+      root = hit::parse(fname, input);
+    }
+    catch (std::exception & err)
+    {
+      std::cerr << err.what() << "\n";
+      return 1;
+    }
 
-      hit::Node * root = nullptr;
-      try
-      {
-        root = hit::parse(fname, input);
-      }
-      catch (std::exception & err)
-      {
-        std::cout << err.what() << "\n";
-        ret = 1;
-        continue;
-      }
+    auto n = root->find(src);
+    if (n)
+    {
+      auto f = dynamic_cast<hit::Field *>(n);
+      delete n;
 
-      DupSectionWalker w;
-      root->walk(&w, hit::NodeType::Section);
+      auto dstnode = new hit::Field(dst, f->kind(), f->val());
+      auto dstnodetree = new hit::Section("");
+      dstnodetree->addChild(dstnode);
+      hit::explode(dstnodetree);
+      hit::merge(dstnodetree, root);
 
-      for (auto & it : w.duplicates)
+      std::string updated = root->render();
+      if (!flags.have("i"))
+        std::cout << updated << "\n";
+      else
       {
-        for (auto sec : it.second)
-        {
-          ret = 1;
-          std::cout << fname << ":" << sec->line() << ": duplicate section name '" << it.first
-                    << "'\n";
-        }
+        std::ofstream f(fname);
+        f << updated;
       }
     }
-    return ret;
   }
-  else
+
+  return ret;
+}
+
+int
+findParam(int argc, char ** argv)
+{
+  Flags flags;
+  flags.add("f", "only show file name");
+  auto positional = parseOpts(argc, argv, flags);
+
+  if (positional.size() < 2)
   {
-    std::cout << "please pass in an input file argument\n";
+    std::cerr << "find expects 2 or more arguments: [parameter-path] [file]...";
     return 1;
   }
+
+  std::string srcpath(positional[0]);
+
+  int ret = 0;
+  for (int i = 1; i < positional.size(); i++)
```

Comment:
signed vs unsigned comparison. These show up as warning in MOOSE compiles and look ugly (our combination of compiler flags). You should avoid mixing of signed types in your project. It'd be nice to have a warning free contrib for once. BTW the proper type isn't `unsigned int` either, it's `std::size_t` for standard containers.

Reply:
Sure, but FWIW this doesn't show up when compiling anything related to the parser for MOOSE - it is a separate binary only compiled if you run the makefile in the hit directory.

Reply:
Clang doesn't seem to warn. Try GCC debug build. It's bad practice anyway.

Reply:
I meant that you won't see this when compiling MOOSE because the main.cc is never compiled as part of anything in MOOSE.